### PR TITLE
Add support for multiple readers

### LINKED ISSUE #2398: Support multiple MetricReaders
We currently only allow one MetricReader (i.e one exporter allowed at same time). Spec requires adding any number of readers/exporters.

Opening this issue to track adding the multi-reader support. (Internally we may simply duplicate the whole pipeline for each reader.)

Treating this as lower priority compared to other features - Views., as most practical scenarios call for one exporter at a time.

### PR DESCRIPTION
**Pull request by utpilla**
Fixes #2398 

## Changes
- Add support for multiple readers

Please provide a brief description of the changes here.

For significant contributions please make sure you have completed the following items:

* [x] `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

File: src/OpenTelemetry/ProviderExtensions.cs
Comment (utpilla) [inline src/OpenTelemetry/ProviderExtensions.cs:53]: This is changing the public API surface but this was only being called by `MetricReader.OnCollect`.

**Comment by reyang** [comment]
Is there a benchmark result?

**Comment by utpilla** [comment]
> Is there a benchmark result?


// * Summary *

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.19043.1348 (21H1/May2021Update)
Intel Core i7-9700 CPU 3.00GHz, 1 CPU, 8 logical and 8 physical cores
.NET SDK=6.0.100
  [Host]     : .NET Core 3.1.21 (CoreCLR 4.700.21.51404, CoreFX 4.700.21.51508), X64 RyuJIT
  DefaultJob : .NET Core 3.1.21 (CoreCLR 4.700.21.51404, CoreFX 4.700.21.51508), X64 RyuJIT

**MetricsBenchmarks original:**

|                    Method | AggregationTemporality |      Mean |    Error |   StdDev | Allocated |
|-------------------------- |----------------------- |----------:|---------:|---------:|----------:|
|            CounterHotPath |             Cumulative |  36.96 ns | 0.083 ns | 0.078 ns |         - |
| CounterWith1LabelsHotPath |             Cumulative | 100.17 ns | 0.118 ns | 0.104 ns |         - |
| CounterWith3LabelsHotPath |             Cumulative | 443.49 ns | 1.101 ns | 1.030 ns |         - |
| CounterWith5LabelsHotPath |             Cumulative | 630.47 ns | 0.951 ns | 0.794 ns |         - |
| CounterWith6LabelsHotPath |             Cumulative | 736.86 ns | 4.960 ns | 4.639 ns |         - |
| CounterWith7LabelsHotPath |             Cumulative | 869.27 ns | 3.129 ns | 2.927 ns |         - |
|            CounterHotPath |                  Delta |  35.97 ns | 0.110 ns | 0.092 ns |         - |
| CounterWith1LabelsHotPath |                  Delta |  99.77 ns | 0.211 ns | 0.187 ns |         - |
| CounterWith3LabelsHotPath |                  Delta | 429.79 ns | 0.616 ns | 0.514 ns |         - |
| CounterWith5LabelsHotPath |                  Delta | 641.03 ns | 1.080 ns | 0.958 ns |         - |
| CounterWith6LabelsHotPath |                  Delta | 748.37 ns | 1.739 ns | 1.627 ns |         - |
| CounterWith7LabelsHotPath |                  Delta | 841.27 ns | 2.278 ns | 2.131 ns |         - |

**MetricsBenchmarks with this PR:**

|                    Method | AggregationTemporality |      Mean |     Error |    StdDev | Allocated |
|-------------------------- |----------------------- |----------:|----------:|----------:|----------:|
|            CounterHotPath |             Cumulative |  37.14 ns |  0.047 ns |  0.044 ns |         - |
| CounterWith1LabelsHotPath |             Cumulative | 103.10 ns |  0.123 ns |  0.115 ns |         - |
| CounterWith3LabelsHotPath |             Cumulative | 434.93 ns |  1.645 ns |  1.539 ns |         - |
| CounterWith5LabelsHotPath |             Cumulative | 639.97 ns |  6.215 ns |  5.814 ns |         - |
| CounterWith6LabelsHotPath |             Cumulative | 744.34 ns |  5.044 ns |  4.212 ns |         - |
| CounterWith7LabelsHotPath |             Cumulative | 854.73 ns | 16.145 ns | 19.827 ns |         - |
|            CounterHotPath |                  Delta |  37.19 ns |  0.043 ns |  0.040 ns |         - |
| CounterWith1LabelsHotPath |                  Delta |  99.42 ns |  0.136 ns |  0.120 ns |         - |
| CounterWith3LabelsHotPath |                  Delta | 440.10 ns |  3.166 ns |  2.961 ns |         - |
| CounterWith5LabelsHotPath |                  Delta | 633.21 ns |  1.330 ns |  1.039 ns |         - |
| CounterWith6LabelsHotPath |                  Delta | 735.38 ns |  1.419 ns |  1.327 ns |         - |
| CounterWith7LabelsHotPath |                  Delta | 841.17 ns |  1.895 ns |  1.680 ns |         - |

**MetricCollectBenchmarks original:**

|  Method | UseWithRef |     Mean |    Error |   StdDev |  Gen 0 | Allocated |
|-------- |----------- |---------:|---------:|---------:|-------:|----------:|
| Collect |      False | 34.93 us | 0.127 us | 0.119 us |      - |     200 B |
| Collect |       True | 24.03 us | 0.095 us | 0.089 us | 0.0305 |     200 B |

**MetricCollectBenchmarks with this PR:**

|  Method | UseWithRef |     Mean |    Error |   StdDev |  Gen 0 | Allocated |
|-------- |----------- |---------:|---------:|---------:|-------:|----------:|
| Collect |      False | 35.17 us | 0.215 us | 0.191 us |      - |     200 B |
| Collect |       True | 23.24 us | 0.028 us | 0.025 us | 0.0305 |     200 B |

**MetricViewBenchmarks original:**

|                                   Method |        ViewConfig |     Mean |   Error |  StdDev | Allocated |
|----------------------------------------- |------------------ |---------:|--------:|--------:|----------:|
| CounterMeasurementRecordingWithThreeTags |            NoView | 468.3 ns | 1.91 ns | 1.78 ns |         - |
| CounterMeasurementRecordingWithThreeTags | ViewNoInstrSelect | 463.0 ns | 2.52 ns | 2.24 ns |         - |
| CounterMeasurementRecordingWithThreeTags |  ViewSelectsInstr | 573.3 ns | 1.29 ns | 1.20 ns |         - |

**MetricViewBenchmarks with this PR:**

|                                   Method |        ViewConfig |     Mean |    Error |   StdDev | Allocated |
|----------------------------------------- |------------------ |---------:|---------:|---------:|----------:|
| CounterMeasurementRecordingWithThreeTags |            NoView | 65.07 ns | 0.112 ns | 0.094 ns |         - |
| CounterMeasurementRecordingWithThreeTags | ViewNoInstrSelect | 65.25 ns | 0.092 ns | 0.086 ns |         - |
| CounterMeasurementRecordingWithThreeTags |  ViewSelectsInstr | 64.91 ns | 0.132 ns | 0.124 ns |         - |

The numbers are significantly low for MetricViewBenchmarks as the benchmark code does not have any reader. With this PR, we do not enable measurements if there is no reader added to the `MeterProviderSdk`

File: src/OpenTelemetry/Metrics/MetricReader.cs
Comment (reyang) [inline src/OpenTelemetry/Metrics/MetricReader.cs:212]: These newly added methods seem to be a good fit for `MetricOwner`/`MetricWriter` 😆

Comment (utpilla) [reply src/OpenTelemetry/Metrics/MetricReader.cs:212]: `MetricReader` does own the `Metric[]` now.

Comment (utpilla) [reply src/OpenTelemetry/Metrics/MetricReader.cs]: I have split the `MetricReader` class into partial classes.

File: src/OpenTelemetry/Metrics/CompositeMetricReader.cs
Comment (reyang) [inline src/OpenTelemetry/Metrics/CompositeMetricReader.cs:158]: Consider putting these in a separate file, too (and later we might consider moving them to a separate class).

**Review by reyang — APPROVED** [review]
LGTM with some suggestions (doesn't have to block this PR).

File: src/OpenTelemetry/Metrics/MeterProviderSdk.cs
Comment (CodeBlanch) [inline src/OpenTelemetry/Metrics/MeterProviderSdk.cs:71]: nit: You don't really need the `is` check. I think you could just execute this `as` line and it ends up `null` if it isn't a `CompositeMetricReader`. Or...

```csharp
if (this.reader is CompositeMetricReader compositeMetricReader)
   this.compositeMetricReader = compositeMetricReader;
```

File: src/OpenTelemetry/Metrics/MeterProviderSdk.cs
Comment (CodeBlanch) [inline src/OpenTelemetry/Metrics/MeterProviderSdk.cs:136]: Is `reader == null` a valid case? Thinking maybe ctor should throw in that case so we don't need to check it.

File: src/OpenTelemetry/ProviderExtensions.cs
Comment (CodeBlanch) [reply src/OpenTelemetry/ProviderExtensions.cs:53]: The name confused me a first because it doesn't return the instruments, it returns a collection callback delegate. A few suggestions...

`GetObservableInstrumentCollect`
`GetObservableInstrumentCollectAction`
`GetObservableInstrumentCollectCallback`

Also looking at the diff it seems we only call this from `MetricReader` could we make it `internal`?

File: src/OpenTelemetry/Metrics/MeterProviderSdk.cs
Comment (reyang) [reply src/OpenTelemetry/Metrics/MeterProviderSdk.cs:136]: Maybe it's fine. Similar discussion here https://github.com/open-telemetry/opentelemetry-dotnet/pull/2476.

File: src/OpenTelemetry/Metrics/MeterProviderSdk.cs
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/MeterProviderSdk.cs:36]: We need to dispose this as well?

File: src/OpenTelemetry/Metrics/MetricReaderExt.cs
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/MetricReaderExt.cs:38]: Isn't this check is done already by the MeterProviderSdk?

Comment (utpilla) [reply src/OpenTelemetry/Metrics/MetricReaderExt.cs:38]: It's only done in the `MeterProviderSdk` when there are no views as we just have to look at `Instrument.Name`.

File: src/OpenTelemetry/Metrics/CompositeMetricReader.cs
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/CompositeMetricReader.cs:72]: we can create the List with the initial size as the number of readers.

File: src/OpenTelemetry/Metrics/CompositeMetricReader.cs
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/CompositeMetricReader.cs:76]: add only if notnull

Comment (cijothomas) [reply src/OpenTelemetry/Metrics/CompositeMetricReader.cs:76]: and in the provider, you can check the list.Count > 0 before enable measurement event is called.

File: src/OpenTelemetry/Metrics/CompositeMetricReader.cs
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/CompositeMetricReader.cs:73]: for a followup/todo - see if we need to catch exceptions from individual readers, so that anyone of them throwing won't affect the others.

File: src/OpenTelemetry/Metrics/MeterProviderSdk.cs
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/MeterProviderSdk.cs:207]: we need to ensure that the MeasuementCompleted callback is expecting List<Metric> here.

File: src/OpenTelemetry/Metrics/MeterProviderSdk.cs
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/MeterProviderSdk.cs:329]: the name of the method no longer makes sense as it handles single stream(Metric) and multiple (List<metric>).
We should make 3 separate methods for Metric, List of Metic, List of (list of metric). Might save all the if-checks in hot path.
Good for a follow up to avoid making this PR bigger.

File: src/OpenTelemetry/ProviderExtensions.cs
Comment (cijothomas) [reply src/OpenTelemetry/ProviderExtensions.cs:53]: `GetObservableInstrumentCollectCallback`

File: src/OpenTelemetry/Metrics/MetricReaderExt.cs
Comment (cijothomas) [reply src/OpenTelemetry/Metrics/MetricReaderExt.cs:38]: yes! But this is the case with no views right?

File: src/OpenTelemetry/Metrics/CompositeMetricReader.cs
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/CompositeMetricReader.cs:120]: only add if the metric.Count > 0

**Review by cijothomas — COMMENTED** [review]
Overall looks good. Left few comments.
Will do another round of review later today before merge

File: src/OpenTelemetry/Metrics/MetricReaderExt.cs
Comment (utpilla) [reply src/OpenTelemetry/Metrics/MetricReaderExt.cs:38]: Whoops! I'm not sure what I did wrong here. I remember undoing this commit last night where I have `IsValidInstrumentName` check happening again.

File: src/OpenTelemetry/Metrics/CompositeMetricReader.cs
Comment (utpilla) [reply src/OpenTelemetry/Metrics/CompositeMetricReader.cs:73]: We don't expect this method to throw any exceptions, right?

File: src/OpenTelemetry/Metrics/CompositeMetricReader.cs
Comment (utpilla) [reply src/OpenTelemetry/Metrics/CompositeMetricReader.cs]: I'm adding `null` to preserve the order in which we process the metrics when we have multiple readers. `CompositeMetricReader` checks for `null` before calling the reader to update the value.

In future, if we allow readers to have their own `MaxMetrics` limit, this would still work as we will know which readers need to update the value and which don't.

File: src/OpenTelemetry/Metrics/CompositeMetricReader.cs
Comment (utpilla) [reply src/OpenTelemetry/Metrics/CompositeMetricReader.cs]: Same as https://github.com/open-telemetry/opentelemetry-dotnet/pull/2596#discussion_r752685144

File: src/OpenTelemetry/Metrics/MeterProviderSdk.cs
Comment (utpilla) [reply src/OpenTelemetry/Metrics/MeterProviderSdk.cs:204]: Fixed this.

File: src/OpenTelemetry/Metrics/MeterProviderSdk.cs
Comment (utpilla) [reply src/OpenTelemetry/Metrics/MeterProviderSdk.cs:36]: We have a `this.reader?.Dispose()` which should take care of it. I have defensively added a `this.compositeReader.Dispose()` as well now.

File: src/OpenTelemetry/Metrics/MetricReaderExt.cs
Comment (utpilla) [reply src/OpenTelemetry/Metrics/MetricReaderExt.cs]: Fixed this.

**Review by cijothomas — APPROVED** [review]
merging to make progress and pave way for others PR, potentially with merge conflicts!. Have few non-blocking items for follow up.

File: src/OpenTelemetry/Metrics/CompositeMetricReader.cs
Comment (cijothomas) [reply src/OpenTelemetry/Metrics/CompositeMetricReader.cs]: not sure. As a general principle, one reader's bad behaviour shouldn't affect other readers

**Pull request merged**
